### PR TITLE
[Epic] New HTTP Adaptor & Utils

### LINKED ISSUE #316: http: rewrite to use new helper functions
With the introduction of new http helpers based on `unidici`, we need to re-write `language-http` to use the new helpers.

It should also use the new `expandReferences` utility function.

At the time of writing, we will need to handle `expandReferences` if `options.body` is a stream. I think is it reasonable to add logic for this in the new `expandReferences` utility (ie, don't recurse into a stream object, which we should be able to duck type)

This is a major version bump as the `axios` export and `axios` APIs are being removed.

I would like to change the http signatures as little as possible I think - so we should still have `get(path, params, callback)`.

We will have to  go through `mapToAxiosConfig` and decide which properties to maintain and which to drop. I would prefer to support a much smaller API.

It may be helpful to review existing http job code to get a sense of how difficult migration is going to be.

### PR DESCRIPTION
**Pull request by josephjclark**
Epic PR for all the http and common work.

Closes #316 #320 #304

# Changes

* [x] New http helpers in common
* [x] New http adaptor
* [x] deprecated old common.http
* [x] update template to showcase new pattern
* [x] check and maybe improve docs
* [ ] ~~remove axios, superagent and request across the repo~~
* [x] Add linting for undeclared variables (and fix a whole bunch of stuff)

# Overview

This PR has the broad aim of removing axios, superagent and request in favour of a single HTTP client - `undici`.

But we're not quite doing that yet as it's too painful.

As a starting point,  we have added a new set of http helper functions (functions not operations!) into `common`, which you'll find in `common/util/http`. These are wrappers around `undici`. We intend that all new and existing adaptors migrate onto these utils.

The headline feature for this release is HTTP 6.0, which is the existing HTTP adaptor "rebased" onto the common helpers, with axios removed. This comes with loads of breaking changes, so there's a thorough migration guide in the http changelog. HTTP should now be a very lightweight adaptor which just wraps the common helpers.

The template has been changed to show usage of the new common helpers.

Finally, the docs output has been improved (internal links with upper case characters fixed; magic metadata removed; better documentation for http options added through a typedef) and linting has been strengthened (resulting in many small but critical patches on existing adaptors).

### CODE REVIEW

**Comment by josephjclark** [comment]
The examples on docs.openfn.org are likely to be way out of date now :( Maybe we ought to go through them

File: packages/template/package.json [hunk 1]
```diff
@@ -27,7 +27,7 @@
     "configuration-schema.json"
   ],
   "dependencies": {
-    "@openfn/language-common": "^1.8.1"
+    "@openfn/language-common": "^1.12.0"
```
Comment (josephjclark) [inline packages/template/package.json:30]: TODO!! Template needs its version bumping, I guess?

Or maybe we should forever lock it at 1.0.0?

**Comment by josephjclark** [comment]
@taylordowns2000 what is the purpose of the version and changelog in `template`?

We've previously been bumping the version number when we make a change to template, as well as adding a changelog. Like we would for a normal adaptor.

But this is meaningless. If someone copies and pastes the template - as we suggest they do - they have to delete the changelog and reset the version number. It's a waste of time - and if this gets missed, then the new adaptor is published with a confusing history.

Although it's orthogonal to this branch, can I suggest we:
a) Set `template` to v `1.0.0`
b) Remove the `template` changelog and add a  stub entry for the 1.0.0 release ("intial release" would be fine)

**Comment by josephjclark** [comment]
Ah, two things I haven't done here:

* Allow access to the raw request object (response.raw as a dynamic, non-enumerable property should work and not be serialised)
* Allow access to the raw error (again, error.raw should work in the same way as above)

We could release without these - but I probably ought to commit one more day to it.

**Comment by josephjclark** [comment]
I'd also add that we haven't added a steaming API (although you can request a body as a stream). See #370

This is fine and part of the plan. It should not hold up the release.

**Comment by taylordowns2000** [comment]
> Although it's orthogonal to this branch, can I suggest we:
> a) Set `template` to v `1.0.0`
> b) Remove the `template` changelog and add a stub entry for the 1.0.0 release ("intial release" would be fine)

Yes, I agree with this @josephjclark .

**Comment by josephjclark** [comment]
@mtuchi reports an issue if the URL is `/` or ``. Something to do with how the URL is assembled.

**Comment by josephjclark** [comment]
Added a cheeky fix for #304

**Comment by josephjclark** [comment]
@mtuchi caught an issue where parseXML was broken because expandReferences isn't imported.

This wasn't picked up in unit testing or linting, which is a bit scary.

I don't want to to write a parseXML unit test now, so I have added a linting rule for undeclared variables. This triggers a rash of patch fixes to various adaptors.

File: packages/http/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
+
+  ### New features
+
+  - Error mapping. Provide custom errors to response status codes, eg, Eg:
+    `{errors: {404: "You don't have access"}}`
+  - parseAs option. Overrides the default response type. Eg: `{parseAs: "json"}`
+  - Request logging. All requests are now logged to stdout.
+  - tls options. Pass SSL certificates using the `tls` option on any request.
+
+  ### Breaking changes
```
Comment (josephjclark) [inline packages/http/CHANGELOG.md:17]: More migration notes with @mtuchi

* The `params` and `qs` parameters have been removed. Use`query` instead (it's the same)
* `keepCookie`, which basically just pulls cookies out of the header and writes them tostate.data. But not in a nice way IMO. We've removed support for this. We can reimplement later, with a spec, if anyone needs it.
* `successCodes` is now the error map.
* `transformResponse` is just a callback? We've removed it. Use the callback isntead.
* agentOptions - use the TLS option. `rejectUnauthorised` might still work? It's unclear
* `reponseType` is `parseAs`
* `authorisation` we do this for you in config. If you need to set it explicitly, she the header  (use the helper)

Comment (josephjclark) [reply packages/http/CHANGELOG.md:17]: `data` is now `body`

File: packages/http/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,47 @@
...
+### Migration Guide
+
+- Use `query` instead of `params` or `qs` parameters to a request.
+- Convert succcessCodes to the `errorMap`. Ie, `successCodes: [200, 404]`
+  becomes `errorMap: { 404: false }`
+- Instead of `transformResponse`, use the callback parameter.
+- Convert `agentOptions` to `tls`.The values should be the same.
+- Convert `responseType` to `parseAs`. Note that this is really only useful for
+  JSON, and only needed if the server does not include a content-type response
+  heade.r
```
Comment (mtuchi) [inline packages/http/CHANGELOG.md:36]: typo, should be header

File: packages/http/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,47 @@
...
+- The `authorisation` property is no longer needed, use state.configuration. The
+  `data` paramter is now `body`.
+
```
Comment (mtuchi) [inline packages/http/CHANGELOG.md:39]: I think we should also add 
- Convert `formData` to `form`.The values should be the same.

**Comment by stuartc** [comment]
> If the server returns an error message, we need to include it in the error object

This is great! A lot of people will be very happy. I can't remember why (if any reason) we didn't do this before? I know there has like always been security considerations around this, like someone logging headers that have API keys in them.

Perhaps it's a non-issue now - but perhaps worth asking @taylordowns2000 why we previously cut error responses from HTTP requests to almost nothing.

**Comment by josephjclark** [comment]
@stuartc The error comments are more related to this branch than to main. I think the error reporting used to be mostly fine - the big problem was that the whole response object was logged, leaking all kinds of stuff and spamming the output.

The new adaptor builds an error object and returns hand-picked keys, which is a problem if we return the wrong keys.

I do want to enable access to the raw request, just in case, but there's a separate issue for that. I think we should do it soonish. Part of  the problem is it's hard to test this stuff, and when you have  a test implementation it's tempting to lean into it a bit too hard.

**Review by stuartc — APPROVED** [review]

**Pull request merged**